File: 0322-coin-change/0322-coin-change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        if amount == 0:
            return 0
        
        change = [ float('inf') for i in range(amount + 1)]
        change[0] = 0
        
        for money in range(amount + 1):
            for coin in coins:
                prevAmount= money - coin
                if prevAmount < 0:
                    continue
                change[money] = min(change[money], change[prevAmount] + 1)
        
        return -1 if change[-1] == float('inf') else change[-1]
        

File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        // Either take or dont take
        // If take, table[amount - coin] + 1
        // else, 0
        if (amount == 0) return 0;
        
        int table[amount + 1];
        // memset(table, INT_MAX, sizeof(table));
        for (int i = 1; i <= amount; i++) table[i] = INT_MAX;
        table[0] = 0;
        
        for (int i = 1; i <= amount; i++) {
            for (int coin: coins) {
                if (i >= coin && table[i - coin] != INT_MAX) {
                    table[i] = min(table[i], table[i - coin] + 1);
                }
            }
        }
        
        return table[amount] < INT_MAX? table[amount] : -1;
    }
};